cpp/src/arrow/compute/kernels/scalar_compare.cc

```diff
@@ -54,72 +56,106 @@ struct GreaterEqual {
   }
 };
 
-struct Less {
-  template <typename T>
-  static constexpr bool Call(KernelContext*, const T& left, const T& right) {
-    return left < right;
-  }
-};
-
-struct LessEqual {
-  template <typename T>
-  static constexpr bool Call(KernelContext*, const T& left, const T& right) {
-    return left <= right;
-  }
-};
+// Implement Greater, GreaterEqual by flipping arguments to Less, LessEqual
 
-template <typename InType, typename Op>
-void AddCompare(const std::shared_ptr<DataType>& ty, ScalarFunction* func) {
-  ArrayKernelExec exec = codegen::ScalarBinaryEqualTypes<BooleanType, InType, Op>::Exec;
-  DCHECK_OK(func->AddKernel({ty, ty}, boolean(), exec));
+template <typename Op>
```

Comment:
It would improve readability if we formalized the `Op` concept and named them consistently. Even within this PR these are also called `Generator`s. It's not consistently made clear what functions they're expected to provide. https://issues.apache.org/jira/browse/ARROW-9161

Reply:
Generator and Op are not the same thing in this nomenclature.